datafusion/functions-aggregate/src/first_last.rs

```diff
@@ -179,6 +292,420 @@ impl AggregateUDFImpl for FirstValue {
+                    // This may be caused by the different vec.capacity() values?
+                    self.orderings.last().unwrap(),
+                );
+        }
+
+        self.is_sets.resize(new_size);
+    }
+
+    fn update_state(
+        &mut self,
+        group_idx: usize,
+        orderings: &[ScalarValue],
+        new_val: T::Native,
+        is_null: bool,
+    ) {
+        self.vals[group_idx] = new_val;
+        self.is_sets.set_bit(group_idx, true);
+
+        self.null_builder.set_bit(group_idx, !is_null);
+
+        assert!(orderings.len() == self.ordering_req.len());
+        let old_size = ScalarValue::size_of_vec(&self.orderings[group_idx]);
+        self.orderings[group_idx].clear();
+        self.orderings[group_idx].extend_from_slice(orderings);
+        let new_size = ScalarValue::size_of_vec(&self.orderings[group_idx]);
+        self.size_of_orderings = self.size_of_orderings - old_size + new_size;
+    }
+
+    // should be used in test only
+    #[cfg(test)]
+    fn compute_size_of_orderings(&self) -> usize {
+        self.orderings
+            .iter()
+            .map(ScalarValue::size_of_vec)
+            .sum::<usize>()
+    }
+
+    /// Returns a hashmap where each group (identified by `group_indices`) is mapped to
+    /// the index of its minimum value in `orderings`, based on lexicographical comparison.
+    /// The function filters values using `opt_filter` and `is_set_arr`
+    fn get_filtered_min_of_each_group(
+        &self,
+        orderings: &[ArrayRef],
+        group_indices: &[usize],
+        opt_filter: Option<&BooleanArray>,
+        vals: &PrimitiveArray<T>,
+        is_set_arr: Option<&BooleanArray>,
+    ) -> Result<HashMap<usize, usize>> {
+        let mut result = HashMap::with_capacity(orderings.len()); // group_idx -> idx_in_orderings
+
+        let comparator = {
+            assert_eq!(orderings.len(), self.ordering_req.len());
+            let sort_columns = orderings
+                .iter()
+                .zip(self.ordering_req.iter())
+                .map(|(array, req)| SortColumn {
+                    values: Arc::clone(array),
+                    options: Some(req.options),
+                })
+                .collect::<Vec<_>>();
+
+            LexicographicalComparator::try_new(&sort_columns)?
+        };
+
+        for (idx_in_val, group_idx) in group_indices.iter().enumerate() {
+            let group_idx = *group_idx;
+
+            let passed_filter = opt_filter.is_none_or(|x| x.value(idx_in_val));
+
+            let is_set = is_set_arr.is_none_or(|x| x.value(idx_in_val));
+
+            if !passed_filter || !is_set {
+                continue;
+            }
+
+            if !self.need_update(group_idx) {
+                continue;
+            }
+
+            if self.ignore_nulls && vals.is_null(idx_in_val) {
+                continue;
+            }
+
+            if !result.contains_key(&group_idx)
```

Comment:
This could be optimized to 1 lookup (e.g. using `HashMap::entry`). It looks in your profile this is a hot function.

Reply:
https://github.com/apache/datafusion/pull/15266/commits/699bcac1bb87aeb020e320ed031ad93476a1a618